### PR TITLE
add support for attribute[*] with capture_jmx_metrics

### PR DESCRIPTION
**Pull request by jackshirazi**
## What does this PR do?
add support for attribute[*] with capture_jmx_metrics - not general wildcard in the attribute, but this specific token which matches all attributes for the associated object name

## Checklist

- [x] This is an enhancement of existing features, or a new feature in existing plugins
  - [x] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [x] I have added tests that prove my fix is effective or that my feature works
  - [x] Added an API method or config option? Document in which version this will be introduced
  - [x] I have made corresponding changes to the documentation
- [ ] This is a bugfix
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [ ] I have added tests that would fail without this fix
- [ ] This is a new plugin
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)
  - [ ] My code follows the [style guidelines of this project](https://github.com/elastic/apm-agent-java/blob/main/CONTRIBUTING.md#java-language-formatting-guidelines)
  - [ ] I have made corresponding changes to the documentation
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] New and existing [**unit** tests](https://github.com/elastic/apm-agent-java/blob/main/CONTRIBUTING.md#testing) pass locally with my changes
  - [ ] I have updated [supported-technologies.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/docs/supported-technologies.asciidoc)
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] Added an instrumentation plugin? Describe how you made sure that old, non-supported versions are not instrumented by accident.
- [ ] This is something else
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/main/CHANGELOG.asciidoc)

### CODE REVIEW

File: apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java [hunk 1]
```diff
@@ -30,18 +30,7 @@
...
-import javax.management.MBeanServer;
-import javax.management.MBeanServerDelegate;
-import javax.management.MBeanServerFactory;
-import javax.management.MBeanServerNotification;
-import javax.management.MalformedObjectNameException;
-import javax.management.Notification;
-import javax.management.NotificationListener;
-import javax.management.ObjectInstance;
-import javax.management.ObjectName;
+import javax.management.*;
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java:33]: [minor] do not use wildcards for imports (this setting broke recently in my IDE too)

File: apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java [hunk 2]
```diff
@@ -308,40 +297,77 @@ private static void addJmxMetricRegistration(final JmxMetric jmxMetric, List<Jmx
...
+                            key,
+                            objectName));
+                } else {
+                    if (!"*".equals(attribute.getJmxAttributeName())) {
+                        logger.warn("Can't create metric '{}' because composite value '{}' is not a number: '{}'", jmxMetric, key, value);
+                    }
+                }
+            }
+        } else {
+            if (!"*".equals(attribute.getJmxAttributeName())) {
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java:365]: [minor] we could probably factorize this by adding `isWildCard` to `JmxMetric.Attribute` class for the 3 instances of this expression.

File: apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java [hunk 3]
```diff
@@ -308,40 +297,77 @@ private static void addJmxMetricRegistration(final JmxMetric jmxMetric, List<Jmx
...
+                return labels.getValue(i)+".";
+            }
+        }
+        return "";
+    }
+
+    private static void addJmxMetricRegistration(JmxMetric jmxMetric, List<JmxMetricRegistration> registrations, ObjectName objectName, Object value, JmxMetric.Attribute attribute, String attributeName, String metricPrepend) throws AttributeNotFoundException {
+        if (value instanceof Number) {
+            logger.debug("Found number attribute {}={}", attribute.getJmxAttributeName(), value);
+            registrations.add(new JmxMetricRegistration(attribute.getMetricName(metricPrepend == null? attributeName : metricPrepend+attributeName),
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java:343]: [minor] probably lacks some formatting as there are missing spaces around `+` sign.

File: apm-agent-plugins/apm-jmx-plugin/src/test/java/co/elastic/apm/agent/jmx/JmxMetricTrackerTest.java [hunk 4]
```diff
@@ -83,6 +83,17 @@ void testGC() throws Exception {
         printMetricSets();
     }
 
+    @Test
+    void testAttributeWildcard() throws Exception {
+        setConfig(JmxMetric.valueOf("object_name[java.lang:type=GarbageCollector,name=*] attribute[*]"));
+        for (GarbageCollectorMXBean gcBean : ManagementFactory.getGarbageCollectorMXBeans()) {
+            String memoryManagerName = gcBean.getName();
+            assertThat(metricRegistry.getGaugeValue("jvm.jmx.collection_count", Labels.Mutable.of("name", memoryManagerName).add("type", "GarbageCollector"))).isNotNegative();
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-jmx-plugin/src/test/java/co/elastic/apm/agent/jmx/JmxMetricTrackerTest.java:91]: This means that the internal JVM metrics are altered by the JMX configuration, which is not something I would have expected, but also probably helps to reproduce the issue that someone reported in our forum a few days ago, I'll investigate this further.

**Review by SylvainJuge — DISMISSED** [review]

File: apm-agent-plugins/apm-jmx-plugin/src/test/java/co/elastic/apm/agent/jmx/JmxMetricTrackerTest.java (hunk 4 above)
Comment (SylvainJuge) [reply apm-agent-plugins/apm-jmx-plugin/src/test/java/co/elastic/apm/agent/jmx/JmxMetricTrackerTest.java:91]: I was too quick to conclude on this, it has the `jvm.jmx.` prefix here, thus while the metrics look similar to the default JVM metrics, those are not the same.

File: apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java (hunk 3 above)
Comment (jackshirazi) [reply apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java]: thanks, reformatted

File: apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java (hunk 2 above)
Comment (jackshirazi) [reply apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java]: thanks, yes config lost after a reset. Fixed in the commit

File: apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java (hunk 1 above)
Comment (jackshirazi) [reply apm-agent-plugins/apm-jmx-plugin/src/main/java/co/elastic/apm/agent/jmx/JmxMetricTracker.java]: thanks, yes config lost after a reset. Fixed in the commit

**Review by SylvainJuge — APPROVED** [review]

**Pull request merged**